### ISSUE TITLE
Using global palette

### ISSUE THREAD

**Issue by ladlon** [turtletooth/tahoma2d#217]
I'm pretty confused about how to use the Global palette, especially in Tahoma.  The manual seems to touch mostly on the Studio palette, and it seems the online videos all use the Level palettes exclusively.  I've tinkered around in Tahoma, bringing up the Studio palette window (which seems to contain the Global palette), but I have no clue how to use it, nor even copy/create styles into it.

The Levels palettes work great, but obviously there are styles that are shared by several pieces (ex. skin on a character, or the overal look of the outline around everything), so I'm trying to sort out how to use a single shared style for those, rather than having to change it for every level.. which is what I figured the Global palette is for...  but I just can't seem to figure it out.

Any good vids/docs on that?  Most vids that DO explain the palette system usually are using a single level, and so they never run into the issue of a style shared by multiple levels (as opposed to multiple shapes within a single level.

I tried copying a style from one level to another, but as I figured, it just makes a copy, not an instance (which updates its settings based on it's parent)... so, I'm not sure how to do this.

**Comment by WolfInABowl** [comment]
I'm not particularly well versed when it comes to the Studio Palette system but here's how it works as I understand it.
With a level selected, you can right-click on a palette e.g. `basics` and choose `Load into Current Palette`, your level should now have the same styles as that palette. 
Your level's styles will have a white box in the upper right-hand corner indicating that they are a reference to a style in a Studio Palette. You can right-click on any of those styles and choose `Toggle Link to Studio Palette`, this will automatically update those styles when you save Studio Palette they are linked to.


**Something else to do with styles:**
This only applies at render time, but you can use the `External Palette` FX node to apply the palette of an existing level to another level. 
Additionally, something not many people seem to know about is that you can drag a palette to the XSheet/Timeline to create a palette level. You can also use this palette level with the `External Palette` node as shown here:
![image](https://user-images.githubusercontent.com/42905003/94300578-1f48d980-ff61-11ea-83f7-b822801db2a0.png)
The styles are applied on a per-index basis and you are not currently able to remap styles to different indices. 

Additionally, if a style is missing in the palette you are copying but is present in the target, then it will simply appear as black.
For example, if you have a character with 10 styles but you are providing the `External Palette` node with a palette that only has 8, the areas on the character where those 2 missing styles are used will be black.

You can use the `External Palette` node in conjunction with the `Texture` node in order to choose which styles will be replaced and which will remain the same, this can be used to alleviate the issue mentioned above.
Here's the node graph for that:
![image](https://user-images.githubusercontent.com/42905003/94301187-10165b80-ff62-11ea-8e98-740d2d7dda65.png)
And here's the settings for the Texture node, in this example styles 1 through 4 will be replaced with the provided palette and the rest will be unaffected.

**Comment by ladlon** [comment]
You're dominating the forums today, WolfInABowl!

Thanks for that.  The ExternalPalette node seems pretty cool.

So, you haven't found any reference (video or docs) outlining the Global palettes either?

How are people dealing with changing something like the black outline around the characters/objects?  That's a typical style that is used by every level (usually colour #2 (...or 1, if you are starting at zero).... yet each level has its own copy (which isn't an instance)... so, if you wanted to change the outline style of a scene, it seems you'd have to copy and paste the style into every level you have manually, every time you wanted to change/adjust some aspect of the outline... unless you're using some sort of global/shared colour, which obviously lead my attention to the Global palette.

I get the impression that the Global palette is supposed to be just another palette, but you define/store in there styles that you know will be used in more than one level (ex. skin tone, etc)... which it very well might be... but in order for that to be of use, you'd think that the Global palette would be accessible in an existing window which is not updated/affected by selecting a different level (like the Level-based palettes do).  It would be a palette that remains visible/on, and therefore be available/accessible regardless of what level you have selected.

Obviously, there's some way to utilize it, but it really doesn't seem obvious to me, and I can't seem to find any clues.

Meanwhile, I get the impression that the Studio pallet is sort of a next level up, where it's a library of many different palettes (ex. that have been created over the course of several projects), that can be globally accessed by (say) anyone within an animation company... like a shared library of approved character palettes to continually use/reuse for each episode.

But, it seems overkill to try and use the Studio palette... at least for my purpses, and even so, I still don't know how to utilize it (...aside from how you are explaining it, which initially seems more like a post-animation type of 'palette swap', rather than having the palette available to you BEFORE you start drawing/painting.... unless I'm misunderstanding (...which is highly likely right now).

I feel like I'm missing some obvious/critical bit of understanding here, since people obviously are having to deal with changing styles that are used by more than one level (such as the black outline of all shapes.... at least as the default).

**Comment by artisteacher** [comment]
I haven't spent much time with the Studio Palette myself.  I always expect changes in the Studio Palette to apply to the levels using the global or project palettes but that doesn't happen on my end. Toggling the link to the studio palette or updating the color from the Studio Palette (both from the style context menu) seem to have no visible effect. Maybe I'm missing something too?

Alternatively, there's a helpful new feature (#95) that hasn't made it into the manual yet:
<img width="343" alt="Screen Shot 2020-09-25 at 5 16 55 PM" src="https://user-images.githubusercontent.com/21159116/94316794-076a5900-ff53-11ea-9b3a-0283b237f02e.png">
<img width="269" alt="Screen Shot 2020-09-25 at 5 18 57 PM" src="https://user-images.githubusercontent.com/21159116/94316913-41d3f600-ff53-11ea-97e1-6ce58e30969b.png">

1. Create a level (ie. smart raster) (or select a cell in an empty column - this can make a palette for the default level type set in preferences)
2. Add all of the styles you want to use as a default on smart raster levels for the time being
3. In the bottom menu for the palette, save a new default palette for the smart raster levels

Any new smart raster levels will use that default palette. The palettes are not linked though, so changes you make to one will not update the others. You can choose to save a different default later on & any new smart raster levels going forward will use the new default. 

If you updated a default palette and want to update the levels using the old palette, you can adjust the existing palettes using the Studio Palette window if you need to:
<img width="501" alt="Screen Shot 2020-09-25 at 5 25 26 PM" src="https://user-images.githubusercontent.com/21159116/94317735-e276e580-ff54-11ea-9255-d4aeb1c1b168.png"> 
This does need to be done individually for each level you want to update.

All 3 level types can have default palettes. The steps are the same.

It's not a linked palette, but it is a huge time saver.

**Comment by ladlon** [comment]
Ya, see, unless I'm missing something, the whole point of a palette (where you can make a change to a style, and it propogates through all of your other frames/work) kind of gets lost, when using more than one level.

On a single level, it's all great, and works as expected...

I'm not even really sure why each level has its own palette, vs having a single palette that you use throughout all the levels, and not just all the frames within a single level.  Sure, it would be nice to also have the ability to put groups of those styles into folders to keep things tidy, and separate (say) the palette for diffrent characters, or props, etc, so you don't just have this sea of styles all in one bay/bin.  But, that's easy enough.

I must be missing something, as obviously everyone who animates must at some point face the task of changing something like the outlines around each character/object, or something that exists in more than one level.

Hopefully someone chimes in about that, as I'm really curious how they deal with that seemingly common/frequently done task.  Unless there's some multi-select feature that I'm not aware of, having to go through each level, clicking on each element, and manually changing the style settings (even if done by a copy/paste method) would be tedious in a large project.

Interesting find with that undocumented feature... but, ya, just like with the copy/paste of a style from one level to another, it SEEMS like a great solution, until you realize that it's an independent copy, and not a linked instance... so you are back to the 'have to change each level's style settings individually) scenario again.

Even if they would have an option (when you copy/paste a style from one level to another) to make a linked/instance copy, then it would be a simple matter of just copying the 'common' styles from the first level, and then pasting it into the new one before starting to draw.  Better still, though, would be a global palette that is always present, and can be accessed/added to easily... which is what it SEEMS they have... I just don't understand how to use it!  Maybe I'll stumble upon the answer by more tinkering, or further reading of the manual... but if you haven't spotted it so far, I'm not terribly confident I'll find anything.

Looks like I'll try another Google search session.  Maybe some unrelated tutorial vid will mention it in passing somewhere.

**Comment by artisteacher** [comment]
I've suggested having the option to link palettes in the past. From what I remember from the discussion, it's a little more complicated than it seems on the code side.  Some palette's are inside the level files themselves, which could potentially cause some backwards compatibility issues.

If you find anything helpful about the global palettes, please share!

**Comment by ladlon** [comment]
True enough... Since things are saved separately, there may be issue with a level's 'reliance' on the 'master palette'... although, I'm not entirely sure why.  But, I'm not a coder, so I'll trust their claim.

But, again, how DOES everyone actually work around the fact that each level has a separate, unlinked palette?  Imagine a scene with 50 different objects (body parts of characters, props, etc)... and midway through, the client says he wants the outlines to be 'brownish' rather than black.  How do you globally change all those?  And seemingly without some convenient 'select all the levels and apply the change all at once' feature?

Yep, if I figure something out, I'll let you know.  So far, it seems that ExternalPalette node might be a solution, although I have to look further into it, as I'm not previously familiar with it.

**Comment by ladlon** [comment]
Hey, artisteacher.  Hopefully you are still following this thread.

Good news, after a LOT of very frustrating trial/error, and fruitless video/thread searching online, I finally think I figured it out.

Quite easy, as I figured... once you actually know how to do it.

So, here's the ladlon version of the explanation, which hopefully will make sense to you (and anyone else following along who is a bit confused by this).  A lot of the online stuff I saw just kind of confused matters!

So, start a project, and open up the Studio Palette window (which is not present in the default Tahoma layouts).
There you will see the file browser with the two types of Studio palettes... Global and Project.
'Global' will be 'visible/presented' (and therefore available) in ANY session/project you may be doing.
'Project' will only show if you are on the matching project.... kind of keeps the library from being a wall of 'every palette for every project I ever did'.
Now, select whichever of the two folders (palette types) you want to use for this palette you are about to make, right click and select New Palette.
Create all the styles you want for that palette (It seems like all you have is the 'make new style' plus icon button... and that once you have made that new style swatch, you have to right click in it and select Style Editor to actually change the colour (as double clicking on the swatch, only allows you to edit the name, and not bring up the colour editor... weird).

So, keep the resulting colour editor window open while you work on all the styles, otherwise you'll have to do the right click thing to bring up the colour editor each time.

Once you finish all the styles (and now have your completed palette), you can select them, copy, and paste them into the LEVEL palette window.  I find it best to select all the LEVEL palette style swatches, and then right click in it and select Paste Name And Colour.  The benefit of this, is that it will overwrite existing styles, as opposed to appending after it (and causing an ofset in your index numbers, as well as having duplicates of your Index1 colour (strokes).

Do that for all your levels.

Now, here's the critical part.  Up to now, that works great for copying a DEAD (unlinked) copy of the 'master colour library' (Studio palette)... but, as before, making any changes in the Studio palette styles (in the STUDIO palette window, not the indidual LEVEL palette window), will result in the Studio palette obviously being altered, but will not propagate into the individual levels, despite being 'copied'.

That is until you do then next critical step.

In the style swatches inside each LEVEL palette window, you'll note (if you did things right) that there is a white square in the top right corner of each swatch.  This indicates it was copied from the Studio palette... but is 'dead' (unlinked).

Simply click in the tiny squares, and an X will appear in them.  NOW they are linked!  (So, that IS actually a thing!)

Do that for all the swatches of all the levels.

Now, to adjust the styles, edit the swatches in the Studio palette window (again, you may have to initially right click on one of the swatches and select Style Editor.  Make the changes.

To propagate it to all the levels, simply select Save Palette using the menu icon at the bottom right of the Studio Palette window.  Immediately, the styles in the Level Palette window (and your viewscreen) should change.

We have a shared, linked single palette!

(BTW, does that 'Save As Default Vector Palette' command set it as the default vector palette for ALL projects, or just the current one??  I was always afraid to click that, as I thought it might overwrite the global (new project) one, as opposed to just for any new scene in the CURRENT project only!)

Hope that helps.  A great weight has been lifted from my shoulders, and I can actually continue with the project now!

(To be totally fair, this is in the manual!  I just somehow missed the link part.  I find the manual is a bit visually busy, and kind of needs some more visual separation/definition between subjects.  Kind of a wall of text, and makes my eyes cross a bit!  Still, it's all there (in the Managing Palettes And Styles part, under Using The Studio Palette.... go figure).  Well, I'll gladly take the shame of not noticing that, and spending days searching the internet without results, if it means I can finally understand the workflow and continue on with my project!)

**Comment by artisteacher** [comment]
Thanks for the description. So this is actually broken on the macOS end. I can see the white boxes in the corner of the swatches, but clicking on them does nothing. Using the context menu to "Toggle Link to Studio Palette" or "Get Color from Studio Palette" both have no impact. Looking in OT, clicking on the white squares doesn't work. The context menu does though.

**Comment by ladlon** [comment]
Well, that's a drag...  You should slam in a bug report of that.  They seem to be very prompt and eager to address bugs here.  They apparently already addressed the 'Tahoma doesn't maximize properly' issue I pointed out a day or so ago.

**Comment by artisteacher** [comment]
I did. @jeremybullock & @manongjohn are both amazing! They are implementing both bug fixes & fantastic new features at a super fast pace. Such a nice change from OT where progress is much, much slower.

**Comment by ladlon** [comment]
OT progress was slow?  Well, I guess I made the right move to switch to Tahoma.

I was using OT for literally a day or so, and then discovered Tahoma by accident.... so, I never fully got the OT experience.

It DID seem like Tahoma's vibe was more 'active' and 'enthusiastic', as far as new features and overall outlook.

**Comment by ladlon** [comment]
Interesting....  I just had the same thing happen to me just now that you had happen in OT.

I could no longer click the link boxes... but luckily I could still use the context menu version.

I wonder what triggered that.

Add that to your bug report thread, if you like.  Seems it's not just a Mac thing afterall!

**Comment by animatewithdermot** [comment]
Years later reply! - this thread was a great help - thanks ladlon ... I was able to figure it out thank to your post.

I made a YT vid showing the process, hopefully this will also be useful.

https://youtu.be/lJZimrkfZXU
